overlord/exportstate/exportstate.go

```diff
@@ -0,0 +1,141 @@
+	}
+	key := stateMapKey(instanceName, rev)
+	if m == nil {
+		delete(exports, key)
+	} else {
+		data, err := json.Marshal(m)
+		if err != nil {
+			panic("internal error: cannot marshal export manifest: " + err.Error())
+		}
+		raw := json.RawMessage(data)
+		exports[key] = &raw
+	}
+	st.Set("exports", exports)
+}
+
+// Get recalls export manifest of a particular snap revision.
+func Get(st *state.State, instanceName string, rev snap.Revision, m *Manifest) error {
+	*m = Manifest{}
+
+	var exports map[string]*json.RawMessage
+	if err := st.Get("exports", &exports); err != nil {
+		// This can return state.ErrNoState
+		return err
+	}
+	key := stateMapKey(instanceName, rev)
+	raw, ok := exports[key]
+	if !ok {
+		return state.ErrNoState
+	}
+	// XXX: do we need the address?
+	if err := json.Unmarshal([]byte(*raw), &m); err != nil {
+		return fmt.Errorf("cannot unmarshal export manifest: %v", err)
+	}
+	return nil
+}
+
+// exporetedVersionSymlinkPath returns the path of the current exported version symlink
+// for given export name.
+func exportedVersionSymlinkPath(exportedName string) string {
+	return filepath.Join(dirs.ExportDir, exportedName, "current")
+}
+
+// updateExportedVersion updates or removes the exported version symlink.
+//
+// If exportedVersion is not empty then it indicates the new version to use.
+// If the symbolic link cannot be created because the export directory does not
+// exist no error is reported. This is because this function is most often
+// called from link-snap, where it runs unconditionally, but most snaps do not
+// have any content to export and the symlink would be dangling.
+//
+// if exportedVersion is empty then then it indicates that no version is exported
+//
+// Appropriate version can be computed by exportedVersionForSnap.
+func updateExportedVersion(exportedName, exportedVersion string) error {
+	pathName := exportedVersionSymlinkPath(exportedName)
+	if exportedVersion != "" {
+		if err := osutil.AtomicSymlink(exportedVersion, pathName); err != nil && !os.IsNotExist(err) {
+			return fmt.Errorf("cannot set exported version of %q to %q: %v", exportedName, exportedVersion, err)
+		}
+	} else {
+		if err := os.Remove(pathName); err != nil && !os.IsNotExist(err) {
+			return fmt.Errorf("cannot unset exported version of %q: %v", exportedName, err)
+		}
+	}
+	return nil
+
+}
+
+// ExportedNameVersion returns the (snapName, exportedVersion) tuple to
+// use as the current provider of all the export sets of a given snap. The
+// returned exportedVersion may be empty, indicating that given snap has no
+// current revision.
+func ExportedNameVersion(st *state.State, instanceName string) (exportedName string, exportedVersion string, err error) {
+	switch instanceName {
+	case "core", "snapd":
+		exportedName = "snapd"
+		exportedVersion, err = effectiveExportedVersionForSnapdOrCore(st)
+		if err != nil {
+			return "", "", err
+		}
+	default:
+		info, err := snapstateCurrentInfo(st, instanceName)
```

Comment:
we should always do this and then for consistency use the type to detect core/snapd

Reply:
This has subtly different meaning. The code as it stands now, will respond with information on "snapd" even if that snap is not installed in the system. We could special case on type if the snap exists but we'd still need the name special case and fall back to what we are doing here now.

Reply:
is the special case ever depended upon though?  exportSnapdTools is calling effectiveExportedVersionForSnapdOrCore directly afaict.   This function is only called by SnapLinkageChanged afaict, which should always mean we have a real snap around? is not even clear it should be exported, maybe is used in follow up PRs ?

Reply:
It's not used in follow-up PRs, I think you are right that it can be less generic and can handle the simple case.